python/plugins/processing/ProcessingPlugin.py

```diff
@@ -282,6 +283,57 @@ def initGui(self):
         self.iface.actionToggleEditing().triggered.connect(partial(self.sync_in_place_button_state, None))
         self.sync_in_place_button_state()
 
+        self.projectProvider = QgsApplication.instance().processingRegistry().providerById("project")
+        self.projectProvider.algorithmsLoaded.connect(self.updateProjectModelMenu)
+        self.projectModelsMenu = None
+        self.projectMenuAction = None
+        self.projectMenuSeparator = None
+
+    def executeProjectModel(self, model, as_batch=False):
+        """Executes a project model"""
+
+        m = self.toolbox.algorithmTree.model()
+
+        # Search for model
+        modelItem = None
+        for r in range(m.rowCount()):
+            i = m.index(r, 0)
+            if m.data(i) == self.projectProvider.name():
+                for rr in range(m.rowCount(i)):
+                    ii = m.index(rr, 0, i)
+                    for rrr in range(m.rowCount(ii)):
+                        modelItem = m.index(rrr, 0, ii)
+                        break
+
+        if modelItem is not None:
+            sm = self.toolbox.algorithmTree.selectionModel()
+            sm.clear()
+            sm.select(modelItem, QItemSelectionModel.Select)
+            if as_batch:
+                self.toolbox.executeAlgorithmAsBatchProcess()
+            else:
+                self.toolbox.executeAlgorithm()
+
+    def updateProjectModelMenu(self):
```

Comment:
A cleaner approach here would be to create the models menu once in the initGui method, and then connect it's menuAboutToShow signal to a slot here which clears the menu, and populates the menu with the current project models.

Reply:
The idea was to not have the models menu item if there are no models, if we create the models in any case it will be empty most of the times.

My proposed implementation works well: by reusing the toolbox I avoided to duplicate the code that handles check and error reporting when the algorithms are launched, this way we have the same behavior of the toolbox when algorithms are launched from the the project->models menu.

But if you insist, I can certainly implement it the other way.

Reply:
> The idea was to not have the models menu item if there are no models, if we create the models in any case it will be empty most of the times.

I think it should always be present -- that follows the same approach as the layouts submenu then, which is always shown even if a project has no layouts. (We could make this a bit nicer and show a single disabled "No Print Layouts" /"No Project Models" entry in the submenu if you want).

I'm fairly certain I've read HIG guidelines which state that it's bad practice for menu entries to appear/disappear in different contexts, and instead they should always appear and be enabled/disabled. This way the user isn't left wondering if they are looking in the wrong place for functionality they are hunting for, and instead are shown immediately that they're looking in the correct place but that functionality isn't available right now. (Can't find the source offhand, but I agree with the logic anyway!)

Reply:
> My proposed implementation works well: by reusing the toolbox I avoided to duplicate the code that handles check and error reporting when the algorithms are launched, this way we have the same behavior of the toolbox when algorithms are launched from the the project->models menu.

> But if you insist, I can certainly implement it the other way.

I'd prefer we didn't use the toolbox approach, yes. There's a couple of reasons:

1.  I'd find it counter-intuitive/annoying that the toolbox selection would change when running an model from the main menu
2. I wouldn't consider any of the python toolbox code as stable api, and there's a ~~good~~ great chance that someone will modify this in future without realising that it's breaking this functionality here.

Reply:
Ok, done.